tests/test_qt/test_qactions.py

```diff
@@ -0,0 +1,56 @@
+from typing import TYPE_CHECKING
+from unittest.mock import Mock
+
+from app_model.backends.qt import QCommandRuleAction, QMenuItemAction
+from app_model.types import Action, MenuItem, ToggleRule
+
+if TYPE_CHECKING:
+    from qtpy.QtWidgets import QAction
+
+    from app_model import Application
+    from conftest import FullApp
+
+
+def test_cache_qaction(qapp, full_app: "FullApp") -> None:
+    action = next(
+        i for k, items in full_app.menus for i in items if isinstance(i, MenuItem)
+    )
+    a1 = QMenuItemAction(action, full_app)
+    a2 = QMenuItemAction(action, full_app)
+    assert a1 is a2
+    assert repr(a1).startswith("QMenuItemAction")
+
+
+def test_toggle_qaction(qapp, simple_app: "Application") -> None:
+    mock = Mock()
+    x = False
+
+    def _connect(qaction: "QAction") -> None:
+        qaction.toggled.connect(mock)
+
+    def _toggle() -> None:
+        nonlocal x
+        x = not x
+
+    action = Action(
+        id="test.toggle",
+        title="Test toggle",
+        toggled=ToggleRule(initialize=lambda: x, experimental_connect=_connect),
+        callback=_toggle,
+    )
+    simple_app.register_action(action)
+
+    a1 = QCommandRuleAction(action, simple_app)
+    assert a1.isCheckable()
+    assert not a1.isChecked()
+
+    a1.trigger()
+    assert a1.isChecked()
+    assert x
+    mock.assert_called_once_with(True)
+    mock.reset_mock()
+
+    a1.trigger()
+    assert not a1.isChecked()
+    assert not x
+    mock.assert_called_once_with(False)
```

Comment:
here I do not see checking if the change of the `x` value from the code will be reflected in the `a1` state.

Reply:
correct, because app-model makes no assumption that `x` is somehow "evented".  The test you're looking for here would be made on the napari side, essentially: to check that changing the value of some evented object x calls `a1.trigger()` ... the corresponding test here would simply be to call `trigger()` directly; that is, we're not testing event systems here.

Reply:
But there should be a such an option like `refresh()` that will evaluate `condition` to check if it should be checked. Then the connection could be made on the side of the backend without exposing this on the side of the definition.

Reply:
but what would trigger the call of refresh?  either we poll periodically (obviously bad) or the consumer (napari) needs to tell us when something appropriate has happened right?  and that would be an event?  (sorry if I'm missing something obvious here!)

Reply:
If we talk about menus, the trigger could be on the open menu (https://doc.qt.io/qt-6/qmenu.html#aboutToShow). 

I see more scenarios, but this require things from `experimental_connect` 

Having such `refresh()` function could allow to remove `experimental_connect` and have something that could be connected to signals.